Packet++/src/IPv4Layer.cpp

```diff
@@ -290,9 +290,15 @@ void IPv4Layer::parseNextLayer()
 	case PACKETPP_IPPROTO_GRE:
 		greVer = GreLayer::getGREVersion(payload, payloadLen);
 		if (greVer == GREv0)
-			m_NextLayer = new GREv0Layer(payload, payloadLen, this, m_Packet);
+		{
+			if (GREv0Layer::isDataValid(payload, payloadLen))
+				m_NextLayer = new GREv0Layer(payload, payloadLen, this, m_Packet);
+		}
 		else if (greVer == GREv1)
-			m_NextLayer = new GREv1Layer(payload, payloadLen, this, m_Packet);
+		{
+			if (GREv1Layer::isDataValid(payload, payloadLen))
+				m_NextLayer = new GREv1Layer(payload, payloadLen, this, m_Packet);
+		}
```

Comment:
It's a bit nicer to do:
```cpp
if (greVer == GREv0 && GREv0Layer::isDataValid(payload, payloadLen))
	m_NextLayer = new GREv0Layer(payload, payloadLen, this, m_Packet);
else if (greVer == GREv1 && GREv1Layer::isDataValid(payload, payloadLen))
	m_NextLayer = new GREv1Layer(payload, payloadLen, this, m_Packet);
else
	m_NextLayer = new PayloadLayer(payload, payloadLen, this, m_Packet);
```

Reply:
But it means if it is `GREv0` and invalid, it will go to `new PayloadLayer`.

Reply:
yes, we always want to have a next layer... the default is `PayloadLayer` if there is no better option...